### PR TITLE
feat(trainer): use seq_lens as the packed-sample boundary contract

### PR DESCRIPTION
**Pull request by hubert-marek**
## Summary

Rebased replacement for #2944, now targeting `main` directly.

- Remove the unused `stack` SFT path and `pack_function`; SFT now has one cat-packed data path with an explicit single-row contract.
- Carry required `seq_lens` through internal SFT/RL batches, transport, dataloaders, and trainer forward. Generic Hugging Face models never receive PrimeRL-only kwargs.
- Derive custom-model FlashAttention `cu_seqlens` only from explicit lengths; public custom-model forwards require keyword-only `seq_lens` and never infer boundaries from `position_ids`.
- Rework SFT cat packing to preserve overflow samples, emit the tail pack, and checkpoint the read-ahead sample so resume cannot skip data.
- Merge trailing padding directly into the final `seq_lens` segment when batches are built, so models receive valid attention lengths without separate padding metadata or forward-time normalization.
- Dense and MoE Qwen3.5 consume boundaries for non-flash linear-attention state resets. Under existing CP, trainers use global lengths for attention setup and derive explicit shard-local lengths for custom model forwards.

## Tests

- GitHub CI: CPU/GPU unit jobs, Ruff, CodeQL, and all integration jobs passed
- Final-stack Slurm H100: reverse-text SFT 4/4, CP-focused 10/10, and RL boundary/packing 35/35 passed; broader model coverage passed 66 tests, with two Qwen3.5 MoE backward tests blocked by the node image missing `cuda/atomic` for TileLang compilation
- Cleanup validation: 30 batch/sequence tests and 22 SFT dataset tests passed; Ruff passed on all changed files
- Direct checkpoint/resume simulation confirms resumed packed output exactly matches uninterrupted execution
- H100 control: resumed steps 11–20 matched uninterrupted training exactly (max loss difference `0.0`); PR and `main` had nearly identical convergence slopes
- Resume integration runs through step 20 so the assertion covers ten post-checkpoint updates
- H100 boundary microbenchmark: explicit `seq_lens` took ~82–83 µs versus ~138–143 µs for position-reset inference across 1K–16K and packed rows

Historical end-to-end validation from #2944:
- Dense and MoE text SFT: 10/10 steps
- RL text run with `seq_lens` end-to-end: 10/10 steps

---

> [!NOTE]
> **High Risk**
> Changes how packed attention boundaries are computed across many custom model forwards and both SFT/RL training paths; incorrect `seq_lens` or CP sharding would silently break causality or loss alignment.
> 
> **Overview**
> Introduces **`seq_lens`** and **`padding_len`** as the canonical packed-row boundary metadata from SFT/RL dataloaders through transport, batch packing, and trainer **`forward`**. Custom PrimeRL models now require keyword-only **`seq_lens`** and build FlashAttention **`cu_seqlens`** via **`get_cu_seqlens_from_seq_lens`** instead of inferring document breaks from **`position_ids`**; several models drop the non-flash causal-mask fallback and always use the varlen path.
> 
> **SFT** removes **`pack_function`** / the **`stack`** packing path and validators tied to it, leaving a single cat-pack pipeline with overflow read-ahead, tail packs, checkpointed pending samples, and collate that expects one packed row per step. **RL** padding extends the last segment’s **`seq_lens`** (and **`padding_len`**) rather than appending a separate padding segment.
> 
> **Context parallelism** passes global **`seq_lens`** into **`setup_cp_params`**, then **`get_cp_local_seq_lens`** for shard-local custom forwards. Generic HF models still do not receive PrimeRL-only kwargs; **`seq_lens`** is injected only for **`PreTrainedModelPrimeRL`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4c77428920ecd5b7d5dd3bd0b1f9b63483bb12b3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/models/afmoe/modeling_afmoe.py [hunk 1]
```diff
@@ -464,6 +464,7 @@ def forward(
         position_ids: Optional[torch.LongTensor] = None,
         inputs_embeds: Optional[torch.FloatTensor] = None,
         routed_experts: Optional[torch.LongTensor] = None,
+        seq_lens: Optional[torch.LongTensor] = None,
```
Comment (samsja) [inline src/prime_rl/trainer/models/afmoe/modeling_afmoe.py]: thos should be optional, lets just say that our custom model needs always seq len

File: src/prime_rl/trainer/model.py [hunk 2]
```diff
@@ -1249,6 +1252,17 @@ def forward(
...
+    if isinstance(model, PreTrainedModelPrimeRL):
+        # Padding is an explicit final segment in transport metadata, but it is
+        # not a document. Merge it into the final real document before model
+        # guards count boundaries; no trainable token follows the padding.
+        model_seq_lens = seq_lens
+        if model_seq_lens is not None and padding_len:
+            if model_seq_lens.numel() < 2:
+                raise ValueError("Padding metadata requires at least one real sequence")
+            model_seq_lens = torch.cat([model_seq_lens[:-2], (model_seq_lens[-2] + model_seq_lens[-1]).view(1)])
+        kwargs["seq_lens"] = model_seq_lens
```
Comment (samsja) [inline src/prime_rl/trainer/model.py]: lest not sure this lets use the config to know if we have custom model or not

File: src/prime_rl/trainer/model.py [hunk 3]
```diff
@@ -1249,6 +1252,17 @@ def forward(
...
 
+    if isinstance(model, PreTrainedModelPrimeRL):
+        # Padding is an explicit final segment in transport metadata, but it is
+        # not a document. Merge it into the final real document before model
+        # guards count boundaries; no trainable token follows the padding.
+        model_seq_lens = seq_lens
+        if model_seq_lens is not None and padding_len:
+            if model_seq_lens.numel() < 2:
+                raise ValueError("Padding metadata requires at least one real sequence")
+            model_seq_lens = torch.cat([model_seq_lens[:-2], (model_seq_lens[-2] + model_seq_lens[-1]).view(1)])
```
Comment (samsja) [inline src/prime_rl/trainer/model.py]: can we create a helper function for this

File: tests/unit/train/test_model_forward.py [hunk 4]
```diff
@@ -18,6 +20,19 @@ def forward(self, **kwargs):
...
 
 
+class _PrimeCaptureModel(PreTrainedModelPrimeRL):
+    config_class = PretrainedConfig
+
+    def __init__(self):
+        super().__init__(PretrainedConfig())
+        self.kwargs = None
+
+    def forward(self, seq_lens: torch.Tensor | None = None, **kwargs):
```
Comment (samsja) [inline tests/unit/train/test_model_forward.py]: hmm lets just add the seq lens to the base class forward and remove this hack

File: tests/unit/test_configs.py [hunk 5]
```diff
@@ -567,6 +567,11 @@ def test_sft_allows_unused_default_renderer_for_fake_data():
     assert config.renderer.name == "default"
 
 
+def test_sft_rejects_removed_pack_function():
+    with pytest.raises(ValidationError, match="pack_function"):
+        SFTConfig.model_validate({"data": {"pack_function": "stack"}})
+
```
Comment (samsja) [inline tests/unit/test_configs.py]: remove tests

File: src/prime_rl/trainer/models/glm4_moe/modeling_glm4_moe.py [hunk 6]
```diff
@@ -205,19 +205,30 @@ def forward(
...
         if (input_ids is None) ^ (inputs_embeds is not None):
             raise ValueError("You must specify exactly one of input_ids or inputs_embeds")
 
         if inputs_embeds is None:
             inputs_embeds: torch.Tensor = self.embed_tokens(input_ids)
 
-        if self.config._attn_implementation in ("flash_attention_2", "flash_attention_3", "fa4"):
-            cu_seqlens, max_seqlen = get_cu_seqlens_from_position_ids(position_ids)
+        flash_attn_enabled = self.config._attn_implementation in ("flash_attention_2", "flash_attention_3", "fa4")
+        if seq_lens.numel() > 1 and not flash_attn_enabled:
```
Comment (S1ro1) [inline src/prime_rl/trainer/models/glm4_moe/modeling_glm4_moe.py]: Im not a fan of these checks, these shouldn't be copy pasted into each model, should be done centrally

File: src/prime_rl/trainer/sft/data.py [hunk 7]
```diff
@@ -331,159 +337,92 @@ def __init__(self, dataset: StatefulIterableDataset, seq_len: int):
...
+            result["position_ids"].extend(range(pad_len))
+            result["loss_mask"].extend([False] * pad_len)
+            result["target_ids"].extend([0] * pad_len)
+            result["seq_lens"][-1] += pad_len
+            result["padding_len"] = pad_len
+        return result
 
 
 def cat_collate(samples: list[Sample]) -> Batch:
+    if len(samples) != 1:
```
Comment (S1ro1) [inline src/prime_rl/trainer/sft/data.py]: this feels useless no?

File: src/prime_rl/trainer/batch.py [hunk 8]
```diff
@@ -491,6 +506,11 @@ def _assert_token_arrays_aligned(micro_batch: MicroBatch) -> None:
     assert sum(micro_batch.sequence_lengths) == num_tokens, (
         f"sequence_lengths sum {sum(micro_batch.sequence_lengths)} != {num_tokens} tokens"
     )
+    assert micro_batch.seq_lens and all(seq_len > 0 for seq_len in micro_batch.seq_lens)
```
Comment (S1ro1) [inline src/prime_rl/trainer/batch.py]: these checks are also quite useless - i.e. this first one, the one on 511/512 too

File: src/prime_rl/utils/sequence.py [hunk 9]
```diff
@@ -50,3 +50,24 @@ def get_cu_seqlens_from_seq_lens(seq_lens: torch.Tensor, total_tokens: int | Non
...
+
+
+def get_cp_local_seq_lens(
+    seq_lens: torch.Tensor,
+    total_tokens: int,
+    cp_rank: int,
+    cp_world_size: int,
+) -> torch.Tensor:
+    """Intersect global sequence boundaries with one contiguous CP shard."""
+    if total_tokens % cp_world_size != 0:
```
Comment (S1ro1) [inline src/prime_rl/utils/sequence.py]: these checks are also validated elsewhere

**Review by S1ro1 — COMMENTED** [review]
lgtm

**Review by samsja — APPROVED** [review]
